Store timestamp of beatmap's last edit time in local database

timiimit opened the pull request:
Closes #23682

This adds `BeatmapInfo.LastEditTime`. `Editor` then reads and uses this variable when it is non-null and assigns it right before beatmap is saved.

timiimit committed:
Add `BeatmapInfo.LastEditTime` property

timiimit committed:
Use `LastEditTime` in `Editor`

bdach commented on osu.Game/Beatmaps/BeatmapInfo.cs:
```diff
@@ -171,6 +171,8 @@ public void ResetOnlineInfo()
 
         public double TimelineZoom { get; set; } = 1.0;
 
+        public double? LastEditTime { get; set; }
```
- Should have xmldoc explaining what this is and where this is used (particularly that this is a user-local setting)
- Since this is a new property persisted to realm, `RealmAccess.schema_version` should be bumped by 1 to reflect the schema change, with an appropriate description in xmldoc therein

peppy commented:
A test would also be appreciated.

peppy committed:
Merge branch 'master' into add-last-edit-time

peppy committed:
Rename variable and ensure timestamp is updated even when not saving

peppy committed:
Bump schema version

peppy committed:
Add test coverage of editor timestamp remembering

peppy commented:
Fixed issues and added test coverage.

timiimit commented on osu.Game/Database/RealmAccess.cs:
```diff
@@ -71,8 +71,9 @@ public class RealmAccess : IDisposable
         /// 24   2022-08-22    Added MaximumStatistics to ScoreInfo.
         /// 25   2022-09-18    Remove skins to add with new naming.
         /// 26   2023-02-05    Added BeatmapHash to ScoreInfo.
+        /// 27   2023-06-06    Added EditorTimestamp to ScoreInfo.
```
```suggestion
        /// 27   2023-06-06    Added EditorTimestamp to BeatmapInfo.
```

peppy committed:
Fix incorrect realm schema comment

Co-authored-by: timiimit <32331609+timiimit@users.noreply.github.com>

bdach committed:
Fix test failures due to cross-test state pollution

`TestSceneEditorTestGameplay` is not isolated from database, and one of
the tests exiting editor when seeked to 60000 milliseconds
(`TestClockTimeTransferIsOneDirectional()`) ended up changing
`EditorTimestamp` to the same value, causing
`TestSaveChangesBeforeGameplayTest()` to fail due to changing initial
state.

To fix, perform a direct deletion of imported beatmaps in realm to avert
this scenario, contrary to the soft-deletion via `BeatmapManager` done
previously.

bdach commented:
I've attempted to fix the test failures via https://github.com/ppy/osu/pull/23711/commits/46ec250d342721c70e54a20795f6994a31c8a12b. Looks fine otherwise, but please do double-check that you're OK with the test fix.

peppy commented:
> I've attempted to fix the test failures via [46ec250](https://github.com/ppy/osu/commit/46ec250d342721c70e54a20795f6994a31c8a12b). Looks fine otherwise, but please do double-check that you're OK with the test fix.

looks good
